calyx/src/passes/comb_prop.rs

```diff
@@ -61,58 +63,88 @@ impl Visitor for CombProp {
             HashMap::new();
 
-        // Build rewrites from unconditional continuous assignments to input
-        // ports of `std_wire`.
-        comp.continuous_assignments
-            .iter()
-            .filter(|assign| {
-                if assign.guard.is_true() {
-                    let dst = assign.dst.borrow();
-                    match &dst.parent {
-                        ir::PortParent::Cell(cell_wref) => {
-                            let cr = cell_wref.upgrade();
-                            let cell = cr.borrow();
-                            cell.is_primitive(Some("std_wire"))
-                        }
-                        ir::PortParent::Group(_) => false,
-                    }
-                } else {
-                    false
+        let parent_is_wire = |parent: &ir::PortParent| -> bool {
+            match parent {
+                ir::PortParent::Cell(cell_wref) => {
+                    let cr = cell_wref.upgrade();
+                    let cell = cr.borrow();
+                    cell.is_primitive(Some("std_wire"))
                 }
-            })
-            .for_each(|assign| {
-                let dst = assign.dst.borrow();
+                ir::PortParent::Group(_) => false,
+            }
+        };
+
+        for assign in &comp.continuous_assignments {
+            // Skip conditional continuous assignments
+            if !assign.guard.is_true() {
+                continue;
+            }
+
+            // src port forwards value on the wire
+            let dst = assign.dst.borrow();
+            if parent_is_wire(&dst.parent) {
                 if let ir::PortParent::Cell(cell_wref) = &dst.parent {
                     let cr = cell_wref.upgrade();
                     let cell = cr.borrow();
                     let dst_idx = cell.get("out").borrow().canonical();
 
                     // If the source has been rewritten, use the rewrite
                     // value from that instead.
-                    let v = port_rewrites
-                        .get(&assign.src.borrow().canonical())
-                        .cloned();
-                    if let Some(pr) = v {
-                        port_rewrites.insert(dst_idx, Rc::clone(&pr));
+                    port_rewrites.insert(dst_idx, Rc::clone(&assign.src));
+                };
+            }
+
+            // wire forwards writes to dst port.
+            let src = assign.src.borrow();
+            if parent_is_wire(&src.parent) {
+                if let ir::PortParent::Cell(cell_wref) = &src.parent {
+                    let cr = cell_wref.upgrade();
+                    let cell = cr.borrow();
+                    let dst_idx = cell.get("in").borrow().canonical();
+                    port_rewrites.insert(dst_idx, Rc::clone(&assign.dst));
+                };
+            }
+        }
+
+        // Make the rewrites consistent.
+        let updates = port_rewrites
```

Comment:
@sampsyo @EclecticGriffin  seems like there should be an obvious data structure that should make this consistency process implicit. Would something like a union-find work here? Not sure if Rust has good union find implementation already.

Reply:
All right, I spent 4 hours trying to make this work with union find by couldn't. Union find represents equality but this needs some form for "directedness". I'm too impatient to learn about anything else right now so going with this instead.